src/java/com/palantir/cassandra/check/VolumesIntegrityCheck.java

```diff
@@ -0,0 +1,125 @@
+
+    public static final String VOLUME_METADATA_NAME = "palantir-startup-metadata.json";
+
+    private final UUID hostId;
+
+    private final FileParser<VolumeMetadata> dataDriveMetadataFileParser;
+
+    private final FileParser<VolumeMetadata> commitLogMetadataFileParser;
+
+    private VolumesIntegrityCheck(UUID hostId)
+    {
+        this(hostId, withParser(getDataDirectory()), withParser(DatabaseDescriptor.getCommitLogLocation()));
+    }
+
+    @VisibleForTesting
+    VolumesIntegrityCheck(UUID hostId,
+                          FileParser<VolumeMetadata> dataDriveMetadataFileParser,
+                          FileParser<VolumeMetadata> commitLogMetadataFileParser)
+    {
+        this.hostId = hostId;
+        this.dataDriveMetadataFileParser = dataDriveMetadataFileParser;
+        this.commitLogMetadataFileParser = commitLogMetadataFileParser;
+    }
+
+    public synchronized void execute()
+    {
+        try
+        {
+            Optional<VolumeMetadata> maybeDataDriveMetadata = dataDriveMetadataFileParser.read();
+            Optional<VolumeMetadata> maybeCommitLogMetadata = commitLogMetadataFileParser.read();
+            check(maybeDataDriveMetadata, maybeCommitLogMetadata);
+
+            /**
+             * Write order matters here based on the implementation for {@link com.palantir.cassandra.check.VolumesIntegrityCheck#check}.
+             * Write to the data drive specifies that we've committed the metadata files to the node. Without writing
+             * to the data drive, we can't differentiate if this is a new node to the cluster or the data drive has
+             * been swapped.
+             */
+            writeIfEmpty(commitLogMetadataFileParser, maybeCommitLogMetadata.isPresent());
+            writeIfEmpty(dataDriveMetadataFileParser, maybeDataDriveMetadata.isPresent());
+        }
+        catch (IOException cause)
+        {
+            throw new RuntimeException("Fail to execute check for volume integrity", cause);
+        }
+    }
+
+    private void check(Optional<VolumeMetadata> maybeDataDriveMetadata, Optional<VolumeMetadata> maybeCommitLogMetadata)
+    {
+        if (maybeCommitLogMetadata.isPresent())
+        {
+            Preconditions.checkState(VolumeMetadata.of(hostId).equals(maybeCommitLogMetadata.get()),
+                                     "Expected SystemKeyspace HostId and POD_NAME to match with commitlog");
+        }
+        else
+        {
+            Preconditions.checkState(!maybeDataDriveMetadata.isPresent(),
+                                     "Expected no metadata has been written to data drive if commitlog is empty");
+        }
+    }
+
+    private void writeIfEmpty(FileParser<VolumeMetadata> parser, boolean present) throws IOException
+    {
+        if (!present) parser.write(VolumeMetadata.of(hostId));
+    }
+
+    public static VolumesIntegrityCheck getInstance(UUID hostId)
+    {
+        if (instance == null) instance = new VolumesIntegrityCheck(hostId);
```

Comment:
Style nit: Eventhough you can remove brackets for "single line if", C* format still puts the action in a new line